sys/linux/io_uring.txt

```diff
@@ -3,43 +3,108 @@
+io_uring_register$IORING_REGISTER_EVENTFD_ASYNC(fd fd_io_uring, opcode const[IORING_REGISTER_EVENTFD_ASYNC], arg ptr[in, fd_event], nr_args const[1])
+io_uring_register$IORING_REGISTER_PROBE(fd fd_io_uring, opcode const[IORING_REGISTER_PROBE], arg ptr[inout, io_uring_probe], nr_args len[arg:ops])
+io_uring_register$IORING_REGISTER_PERSONALITY(fd fd_io_uring, opcode const[IORING_REGISTER_PERSONALITY], arg const[0], nr_args const[0]) ioring_register_personality_id
+io_uring_register$IORING_UNREGISTER_PERSONALITY(fd fd_io_uring, opcode const[IORING_UNREGISTER_PERSONALITY], arg const[0], nr_args ioring_register_personality_id)
 
-mmap$IORING_OFF_SQ_RING(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_io_uring, offset const[IORING_OFF_SQ_RING])
-mmap$IORING_OFF_CQ_RING(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_io_uring, offset const[IORING_OFF_CQ_RING])
-mmap$IORING_OFF_SQES(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_io_uring, offset const[IORING_OFF_SQES])
+mmap$IORING_OFF_SQ_RING(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_io_uring, offset const[IORING_OFF_SQ_RING]) sq_ring_ptr
+mmap$IORING_OFF_CQ_RING(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_io_uring, offset const[IORING_OFF_CQ_RING]) cq_ring_ptr
+mmap$IORING_OFF_SQES(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_io_uring, offset const[IORING_OFF_SQES]) sqes_ptr
 
-io_uring_setup_flags = IORING_SETUP_IOPOLL, IORING_SETUP_SQPOLL, IORING_SETUP_SQ_AFF, IORING_SETUP_CQSIZE
+# If no flags are specificed(0), the io_uring instance is setup for interrupt driven IO.
+io_uring_setup_flags = 0, IORING_SETUP_IOPOLL, IORING_SETUP_SQPOLL, IORING_SETUP_SQ_AFF, IORING_SETUP_CQSIZE, IORING_SETUP_CLAMP
 io_uring_enter_flags = IORING_ENTER_GETEVENTS, IORING_ENTER_SQ_WAKEUP
 _ = __NR_mmap2
 
+# TODO: Consume anything registered within io_uring_sqes as well.
+
+# The ops array in io_uring_probe might contain at most the number available
+# IORING_OP many entries, which corresponds to IORING_OP_LAST. Thus, while
+# probing, let's probe for all and ensure that the kernel can fit all if all of
+# them are supported.
+define IO_URING_PROBE_OPS_SIZE	IORING_OP_LAST
+
+# NEED: part of fields are input here and part are output. We can't express this yet (#245).
+# Only the ops and ops_len is passed in, rest is to be set by kernel
+io_uring_probe {
+	last_op	int8
+	ops_len	len[ops, int8]
+	resv	const[0, int16]
+	resv2	array[const[0, int32], 3]
+	ops	array[io_uring_probe_op, IO_URING_PROBE_OPS_SIZE]
+}
+
+# Values are to be set by kernel
+io_uring_probe_op {
+	op	int8
+	resv	const[0, int8]
+	flags	int16
+	resv2	const[0, int32]
+}
+
+io_uring_files_update {
+	offset	int32
+	resv	int32
+	fds	ptr[in, array[fd]]
```

Comment:
This should be `ptr64`.

Reply:
[fixed](https://github.com/google/syzkaller/pull/1926/commits/4e1574defa49198322992d3381de5d1d634df80a#diff-2e824a90609ef73cc6319dbc406b6b5eR68)